ee/clickhouse/queries/cohort_query.py

```diff
@@ -0,0 +1,476 @@
+from ee.clickhouse.models.cohort import get_count_operator, get_entity_query
+from ee.clickhouse.models.property import prop_filter_json_extract
+from ee.clickhouse.queries.event_query import EnterpriseEventQuery
+from posthog.models import Filter, Team
+from posthog.models.action import Action
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.filters.path_filter import PathFilter
+from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.session_recordings_filter import SessionRecordingsFilter
+from posthog.models.filters.stickiness_filter import StickinessFilter
+from posthog.models.property import Property, PropertyGroup, PropertyName
+from posthog.queries.column_optimizer import ColumnOptimizer
+
+Relative_Date = Tuple[int, str]
+NOW: Relative_Date = (0, "now")
+
+Period = Tuple[Relative_Date, Relative_Date]
+Event = Tuple[str, Tuple[str, int]]
+Event_In_Period = Tuple[Event, Period]
+
+USES_PERIOD = ["performed_event", "performed_event_muliple", "stopped_performing_event", "restarted_performing_event"]
+
+
+class CohortOptimizer(ColumnOptimizer):
+    _events = []
+
+    @cached_property
+    def dates_to_query_column_names(self) -> Dict[Relative_Date, Tuple[str, str]]:
+        _dates = {NOW: "now"}
+        for idx, prop in enumerate(self.filter.property_groups.flat):
+            idx_label = idx * 2  # count by 2 because we can have up to 4 labels per step
+            if prop.time_interval is not None and prop.time_value is not None:
+                _dates[(prop.time_value, prop.time_interval)] = f"date_{idx_label}"
+            if prop.seq_time_interval is not None and prop.seq_time_value is not None:
+                _dates[(prop.seq_time_value, prop.seq_time_interval)] = f"date_{idx_label + 1}"
+        return _dates
+
+    def dates_to_query(self) -> List[Relative_Date]:
+        return list(self.dates_to_query_column_names.keys())
+
+    def get_date_column(self, period: Relative_Date) -> ColumnName:
+        return self.dates_to_query_column_names[period]
+
+    @cached_property
+    def events_in_period_to_query_column_names(self) -> Dict[Event_In_Period, str]:
+        _events = {}
+        for idx, prop in enumerate(self.filter.property_groups.flat):
+            idx_label = idx * 2  # count by 2 because we can have up to 4 labels per step
+            if prop.type in USES_PERIOD:
+                if (
+                    prop.key is not None
+                    and prop.event_type is not None
+                    and prop.time_interval is not None
+                    and prop.time_value is not None
+                ):
+                    _start_date = (prop.time_value, prop.time_interval)
+                    _end_date = NOW
+                    _period = (_start_date, _end_date)
+                    _event = (prop.event_type, prop.key)
+                    self.add_event_type(_event)
+                    _events[
```

Comment:
this data structure isn't making much sense to me right now, judging by how they're being used to get column name based on a period. I guess, question is, why do we have a column like this for each event/period combination?

Reply:
Actually, never mind the second part: That's clever, makes sense now. Instead of an array, it's columns, that can be turned into an array if necessary later on. Makes querying easier. Nice.